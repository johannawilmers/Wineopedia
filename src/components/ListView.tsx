import React, { useState } from "react";
import { Wine } from "../utils/types";
import "../styles/listView.css";
import WineCard from "./WineCard";
import { useFavorites } from "../hooks/useFavorites";

interface WineListView {
  wines: Wine[];
}

/**
 * This component is used to display a list of wine cards
 * @param {WineListView} wines
 */
const ListView: React.FunctionComponent<WineListView> = ({ wines }) => {
  const [visibleItems, setVisibleItems] = useState(6);
  const { favorites, toggleFavorite } = useFavorites();

  const showMore = () => {
    setVisibleItems((prevVisibleItems) => prevVisibleItems + 6);
  };

  return (
    <div>
      <div className="listview-container">
        {wines.slice(0, visibleItems).map((wine) => (
          <div key={wine.basic.productId} className="wine-card-wrapper">
            <WineCard wine={wine} />
            <button
              onClick={() => toggleFavorite(wine.basic?.productId)}
              className="favoriteButton"
            >
              {wine.basic && favorites.includes(wine.basic.productId)
                ? "Remove from Favorites"
                : "Add to Favorites"}
            </button>
          </div>
        ))}
      </div>
      <div id="buttonDiv" data-testid="favoriteButton">
        {visibleItems < wines.length && (
          <button className="show-more-button" onClick={showMore}>
            Load more wines
          </button>
        )}
      </div>
    </div>
  );
};

export default ListView;
