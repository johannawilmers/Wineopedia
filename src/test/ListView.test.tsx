import {
  render,
  screen,
  fireEvent,
  RenderResult,
} from "@testing-library/react";
import { BrowserRouter } from "react-router-dom";
import ListView from "../components/ListView";
import "@testing-library/jest-dom/extend-expect";

const wineArray = [
  {
    basic: {
      productId: "133501",
      productShortName: "Tissot Vin Jaune La Vasée",
      favorite: false,
    },
    lastChanged: {
      date: "2021-04-16",
      time: "06:19:05",
    },
  },
  {
    basic: {
      productId: "162701",
      productShortName: "Niepoort Bioma Vintage",
      favorite: false,
    },
    lastChanged: {
      date: "2021-03-15",
      time: "22:57:42",
    },
  },
];

const wineArrayAfterLoad = [
  {
    basic: {
      productId: "133501",
      productShortName: "Tissot Vin Jaune La Vasée",
      favorite: false,
    },
    lastChanged: {
      date: "2021-04-16",
      time: "06:19:05",
    },
  },
  {
    basic: {
      productId: "162701",
      productShortName: "Niepoort Bioma Vintage",
      favorite: false,
    },
    lastChanged: {
      date: "2021-03-15",
      time: "22:57:42",
    },
  },
  {
    basic: {
      productId: "162000",
      productShortName: "Some new whitewine",
      favorite: false,
    },
    lastChanged: {
      date: "2021-03-15",
      time: "22:57:42",
    },
  },
];

describe("renders ListView component", () => {
  let wrapper: RenderResult;

  beforeEach(() => {
    wrapper = render(
      <BrowserRouter>
        <ListView wines={wineArray} />
      </BrowserRouter>,
    );
  });

  it("create snapshot of listview", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("test that the first wine is visible on site", () => {
    const heading = wrapper.getByRole("heading", {
      name: /Tissot Vin Jaune La Vasée/i,
    });
    expect(heading).toBeInTheDocument();
  });

  it("test that the second wine is visible on site", () => {
    const heading = wrapper.getByRole("heading", {
      name: /Niepoort Bioma Vintage/i,
    });
    expect(heading).toBeInTheDocument();
  });

  it("test that a wine is NOT visible on the site", () => {
    const wineNotInDocument = screen.queryByText("thiswineisnotindocument");
    expect(wineNotInDocument).not.toBeInTheDocument();
  });

  it("should test the number of wines on site", () => {
    expect(wrapper).length == 2;
  });

  it("should load more wines", () => {
    const loadMoreButton = wrapper.getByTestId("favoriteButton");
    fireEvent.click(loadMoreButton);

    wrapper.rerender(
      <BrowserRouter>
        <ListView wines={wineArrayAfterLoad} />
      </BrowserRouter>,
    );

    expect(wrapper.getByText("Some new whitewine")).toBeInTheDocument();
  });

  it("get loadMore button", () => {
    const loadMoreButton = wrapper.getByTestId("favoriteButton");
    expect(loadMoreButton).toBeInTheDocument();
  });
});

describe("Renders emtpy listview", () => {
  it("test if the list is empty", () => {
    render(
      <BrowserRouter>
        <ListView wines={[]} />
      </BrowserRouter>,
    );
    expect(screen.queryByText("Load more wines")).toBeNull();
  });
});
