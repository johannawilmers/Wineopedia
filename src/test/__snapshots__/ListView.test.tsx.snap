// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`renders ListView component > create snapshot of listview 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div>
        <div
          class="listview-container"
        >
          <div
            class="wine-card-wrapper"
          >
            <a
              class="wineCard"
              href="/project1/product/133501"
            >
              <div>
                <img
                  alt="Tissot Vin Jaune La Vasée"
                  src="https://bilder.vinmonopolet.no/cache/300x300-0/133501-1.jpg"
                />
              </div>
              <h2
                class="wineName"
              >
                Tissot Vin Jaune La Vasée
              </h2>
            </a>
            <button
              class="favoriteButton"
            >
              Add to Favorites
            </button>
          </div>
          <div
            class="wine-card-wrapper"
          >
            <a
              class="wineCard"
              href="/project1/product/162701"
            >
              <div>
                <img
                  alt="Niepoort Bioma Vintage"
                  src="https://bilder.vinmonopolet.no/cache/300x300-0/162701-1.jpg"
                />
              </div>
              <h2
                class="wineName"
              >
                Niepoort Bioma Vintage
              </h2>
            </a>
            <button
              class="favoriteButton"
            >
              Add to Favorites
            </button>
          </div>
        </div>
        <div
          data-testid="favoriteButton"
          id="buttonDiv"
        />
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      <div
        class="listview-container"
      >
        <div
          class="wine-card-wrapper"
        >
          <a
            class="wineCard"
            href="/project1/product/133501"
          >
            <div>
              <img
                alt="Tissot Vin Jaune La Vasée"
                src="https://bilder.vinmonopolet.no/cache/300x300-0/133501-1.jpg"
              />
            </div>
            <h2
              class="wineName"
            >
              Tissot Vin Jaune La Vasée
            </h2>
          </a>
          <button
            class="favoriteButton"
          >
            Add to Favorites
          </button>
        </div>
        <div
          class="wine-card-wrapper"
        >
          <a
            class="wineCard"
            href="/project1/product/162701"
          >
            <div>
              <img
                alt="Niepoort Bioma Vintage"
                src="https://bilder.vinmonopolet.no/cache/300x300-0/162701-1.jpg"
              />
            </div>
            <h2
              class="wineName"
            >
              Niepoort Bioma Vintage
            </h2>
          </a>
          <button
            class="favoriteButton"
          >
            Add to Favorites
          </button>
        </div>
      </div>
      <div
        data-testid="favoriteButton"
        id="buttonDiv"
      />
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
